### PR TITLE
Initial Commit.

### PR DESCRIPTION
**Pull request by MichaelNouwens**

### CODE REVIEW

File: handler/ApiHandler.ashx [hunk 1]
```diff
@@ -143,6 +163,37 @@ public class ApiHandler : PluginHandler
...
+        var getParamBuilder = slackClient.GetParams(context.Form["text"].Split(' '), context);
+        dynamic response = JsonConvert.DeserializeObject(slackClient.GetChannelsHistory(context.Form["channel_id"], getParamBuilder.ToString()));
+        List<SlackClient.MessagesFromSlack> getListOfMessages= slackClient.GetAllChat(response);
+        if (!(getListOfMessages == null) && (!getListOfMessages.Any()))
+        {
+            slackClient.SlackMessageTemplate(context, "No Messages were added. Please enter valid parameters, etc: /associate xxxxxxxxxxxxxxxx xxxxxxxxxxxxxxxx");
+        }
+        else
+        {
+
```
Comment (Kgothatso09) [inline handler/ApiHandler.ashx]: Please remove this line

File: handler/ApiHandler.ashx [hunk 2]
```diff
@@ -143,6 +163,37 @@ public class ApiHandler : PluginHandler
...
     }
 
+    private string AssociateFromSlack(HttpRequest context)
+    {
+        var slackClient = new SlackClient(this.BaseUrl, this.SlackAppOAuthToken);
+        var requestNumber = new RequestBroker().FindByNumber(Convert.ToInt32(context.Form["channel_name"].Substring(4)));
+        var getParamBuilder = slackClient.GetParams(context.Form["text"].Split(' '), context);
+        dynamic response = JsonConvert.DeserializeObject(slackClient.GetChannelsHistory(context.Form["channel_id"], getParamBuilder.ToString()));
+        List<SlackClient.MessagesFromSlack> getListOfMessages= slackClient.GetAllChat(response);
+        if (!(getListOfMessages == null) && (!getListOfMessages.Any()))
```
Comment (Kgothatso09) [inline handler/ApiHandler.ashx]: This can be converted to `if (!getListOfMessages?.Any())`

File: handler/ApiHandler.ashx [hunk 3]
```diff
@@ -143,6 +163,37 @@ public class ApiHandler : PluginHandler
         }
     }
 
+    private string AssociateFromSlack(HttpRequest context)
+    {
+        var slackClient = new SlackClient(this.BaseUrl, this.SlackAppOAuthToken);
+        var requestNumber = new RequestBroker().FindByNumber(Convert.ToInt32(context.Form["channel_name"].Substring(4)));
+        var getParamBuilder = slackClient.GetParams(context.Form["text"].Split(' '), context);
+        dynamic response = JsonConvert.DeserializeObject(slackClient.GetChannelsHistory(context.Form["channel_id"], getParamBuilder.ToString()));
+        List<SlackClient.MessagesFromSlack> getListOfMessages= slackClient.GetAllChat(response);
```
Comment (Kgothatso09) [inline handler/ApiHandler.ashx]: Maybe call the variable `messages` =

File: handler/ApiHandler.ashx [hunk 4]
```diff
@@ -143,6 +163,37 @@ public class ApiHandler : PluginHandler
...
+        if (!(getListOfMessages == null) && (!getListOfMessages.Any()))
+        {
+            slackClient.SlackMessageTemplate(context, "No Messages were added. Please enter valid parameters, etc: /associate xxxxxxxxxxxxxxxx xxxxxxxxxxxxxxxx");
+        }
+        else
+        {
+
+            dynamic responseAssociatedUser = JsonConvert.DeserializeObject(slackClient.GetUserInfo(context.Form["user_id"]));
+            var content = slackClient.CreatingNotes(responseAssociatedUser, getListOfMessages);
```
Comment (Kgothatso09) [inline handler/ApiHandler.ashx]: CreateNotes(responseAssociatedUser, getListOfMessages)

File: handler/ApiHandler.ashx [hunk 5]
```diff
@@ -258,6 +309,90 @@ public class ApiHandler : PluginHandler
             return this.CreateMessage(this.baseUri + "chat.postMessage", payloadJson, "POST");
         }
 
+        public string GetChannelsHistory(string channelId, string allParams)
```
Comment (Kgothatso09) [inline handler/ApiHandler.ashx]: maybe the name of the second parameter can be changed to `filters`

File: handler/ApiHandler.ashx [hunk 6]
```diff
@@ -258,6 +309,90 @@ public class ApiHandler : PluginHandler
...
+                        ts = message.ts,
+                        colour = colourForUser.ContainsKey(responseUserRealName.user.id.Value) ? colourForUser[responseUserRealName.user.id.Value] : colourType
+                    });
+                }
+            }
+
+            return listOfMessages;
+        }
+
+        public string CreatingNotes(dynamic response, List<MessagesFromSlack> getListFromMethod)
```
Comment (Kgothatso09) [inline handler/ApiHandler.ashx]: maybe `CreateNotes` or  `AddNotes`?

File: handler/ApiHandler.ashx [hunk 7]
```diff
@@ -258,6 +309,90 @@ public class ApiHandler : PluginHandler
...
+        public string WebhookToSlack(string url, string data, string method)
+        {
+            return this.ProcessRequest(url, data, method);
+        }
+
+        public void SlackMessageTemplate(HttpRequest context, string message)
+        {
+            Dictionary<string, string> data = new Dictionary<string, string>();
+            data.Add("text", message);
+            WebhookToSlack(context.Form["response_url"], JsonConvert.SerializeObject(data), "POST");
```
Comment (Kgothatso09) [inline handler/ApiHandler.ashx]: this.WebhookToSlack()

File: handler/ApiHandler.ashx [hunk 8]
```diff
@@ -27,6 +28,14 @@ public class ApiHandler : PluginHandler
         }
     }
 
+    private string MSMApiKey
```
Comment (Thulasizwe0602) [inline handler/ApiHandler.ashx]: Cant we use expression bodies over here?

File: handler/ApiHandler.ashx [hunk 9]
```diff
@@ -39,11 +48,19 @@ public class ApiHandler : PluginHandler
 
     private string Action { get; set; }
 
+    private string BaseNoteUrl
```
Comment (Thulasizwe0602) [inline handler/ApiHandler.ashx]: Ditto

File: handler/ApiHandler.ashx [hunk 10]
```diff
@@ -39,11 +48,19 @@ public class ApiHandler : PluginHandler
...
 
+    private string BaseNoteUrl
+    {
+        get
+        {
+            return HttpContext.Current.Request.Url.Scheme + "://" + HttpContext.Current.Request.Url.Host + "/MSM/api/serviceDesk/operational/requests/{0}/notes";
+        }
+    }
+
     private string BaseUrl
```
Comment (Thulasizwe0602) [inline handler/ApiHandler.ashx]: ditto

File: handler/ApiHandler.ashx [hunk 11]
```diff
@@ -143,6 +163,37 @@ public class ApiHandler : PluginHandler
...
+        if (!(getListOfMessages == null) && (!getListOfMessages.Any()))
+        {
+            slackClient.SlackMessageTemplate(context, "No Messages were added. Please enter valid parameters, etc: /associate xxxxxxxxxxxxxxxx xxxxxxxxxxxxxxxx");
+        }
+        else
+        {
+
+            dynamic responseAssociatedUser = JsonConvert.DeserializeObject(slackClient.GetUserInfo(context.Form["user_id"]));
+            var content = slackClient.CreatingNotes(responseAssociatedUser, getListOfMessages);
+            Dictionary<string, string> data = new Dictionary<string, string>();
```
Comment (Thulasizwe0602) [inline handler/ApiHandler.ashx]: Lets try using object initialization.

File: handler/ApiHandler.ashx [hunk 12]
```diff
@@ -258,6 +309,90 @@ public class ApiHandler : PluginHandler
...
+        }
+
+        public string WebhookToSlack(string url, string data, string method)
+        {
+            return this.ProcessRequest(url, data, method);
+        }
+
+        public void SlackMessageTemplate(HttpRequest context, string message)
+        {
+            Dictionary<string, string> data = new Dictionary<string, string>();
```
Comment (Thulasizwe0602) [inline handler/ApiHandler.ashx]: Ditto

File: handler/ApiHandler.ashx (hunk 8 above)
Comment (MichaelNouwens) [reply handler/ApiHandler.ashx]: Not compatible with the current version of c#.

File: handler/ApiHandler.ashx (hunk 2 above)
Comment (MichaelNouwens) [reply handler/ApiHandler.ashx]: Not compatible with the current version of c#.

**Pull request merged**